### ISSUE TITLE
A question about computing Ci by k-mean

### ISSUE THREAD

**Issue by MrYaoH** [neuroailab/LocalAggregation-Pytorch#6]
Hi, chengxuz, I have a question. In LA part, the loss backpropogates before k-mean updation. Does this mean that the first data batch did not define Ci as mentioned in the paper？

**Comment by chengxuz** [comment]
At the begining of the training, which is before any loss computation, Ci would be computed using the saved memory bank.

**Comment by MrsYaoH** [comment]
> At the begining of the training, which is before any loss computation, Ci would be computed using the saved memory bank.

In the ‘train_one_epoch‘ function，I notice that when loss.backward() is carried out for the first time，the Ci is not clustered by k-means method. Clustering does not work until the second data batch in the first epoch enters. This problem make me confused.

**Comment by MrsYaoH** [comment]
self.optim.zero_grad()
            loss.backward()
            self.optim.step()
            '''
                why backword before the k-means
            '''
            with torch.no_grad():
                self.memory_bank.update(indices, new_data_memory)
                self.loss_fn.module.memory_bank_broadcast = self.memory_bank.bank_broadcast  # update loss_fn

                if (self.config.loss_params.loss == 'LocalAggregationLossModule' and
                        (self.first_iteration_kmeans or batch_i == self.config.loss_params.kmeans_freq)):

                    if self.first_iteration_kmeans:
                        self.first_iteration_kmeans = False

                    self.logger.info('Fitting K-means with FAISS')

                    # get kmeans clustering (update our saved clustering)
                    k = [self.config.loss_params.kmeans_k for _ in
                         range(self.config.loss_params.n_kmeans)]

                    # NOTE: we use a different gpu for FAISS otherwise cannot fit onto memory
                    self.km = Kmeans(k, self.memory_bank, gpu_device=self.gpu_devices)
                    cluster_labels = self.km.compute_clusters()

                    # broadcast the cluster_labels after modification
                    for i in range(len(self.gpu_devices)):
                        device = self.loss_fn.module.cluster_label_broadcast[i].device
                        self.loss_fn.module.cluster_label_broadcast[i] = cluster_labels.to(device)

**Comment by MrsYaoH** [comment]
I accidentally click the ’closed‘ button，i 

> At the begining of the training, which is before any loss computation, Ci would be computed using the saved memory bank.
